src/System.Linq/src/System/Linq/Concatenate.cs

```diff
@@ -21,18 +21,28 @@ public static IEnumerable<TSource> Concat<TSource>(this IEnumerable<TSource> fir
                 throw Error.ArgumentNull(nameof(second));
             }
 
+            var firstCol = first as ICollection<TSource>;
+            if (firstCol != null)
+            {
+                var secondCol = second as ICollection<TSource>;
+                if (secondCol != null)
+                {
+                    return new Concat2CollectionIterator<TSource>(firstCol, secondCol);
+                }
+            }
```

Comment:
Previously there was only a single type check involved; now there are three.  Your microbenchmark only covered the positive case; what about the negative one?